truffle/com.oracle.truffle.api/src/com/oracle/truffle/api/source/SourceSection.java

```diff
@@ -76,6 +83,30 @@
         this.startColumn = startColumn;
         this.charIndex = charIndex;
         this.charLength = charLength;
+        this.tags = tags;
+        assert tagsAreNonNullAndInterned(tags) : "All tags set for a source section must be interned and non-null.";
+    }
+
+    @SuppressFBWarnings("ES_COMPARING_STRINGS_WITH_EQ")
+    private static boolean tagsAreNonNullAndInterned(String[] tags) {
+        for (int i = 0; i < tags.length; i++) {
+            if (tags[i] == null) {
+                return false;
+            }
+            if (tags[i].intern() != tags[i]) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    /**
+     * Returns a set of tags that are associated with this source section as array. The returned
+     * array must be considered read-only. All elements of the array were asserted to be non-null
+     * and interned for faster comparison.
+     */
+    public String[] getTags() {
```

Comment:
If you want to keep your work experimental, then please stay on a branch or fork truffle repository and play with it. You can research whatever you want in such forks, I don't care.

The primary (e.g. already in production) client of the API is debugger and the current dichotomy of two instrumentation APIs isn't sustainable. I need the API part used by debugger to be merged (now/this week) and I need all languages to adopt it before next Truffle release (by end of February +- two weeks).

The part needed by debugger has to be stable.